### DISCUSSION TITLE
Ahead and behind car gap

### DISCUSSION THREAD

**Discussion by KommyKT** [kaiusl/KLPlugins.DynLeaderboards#21] [Ideas]
I want to do a gap screen for ahead and behind my car based on dynLeaderboard data.
Now I use the following code, this is correct?
This is ahead, behind is idx+2.
`
var idx = $prop('DynLeaderboardsPlugin.Dynamic.FocusedPosInCurrentLeaderboard')
idx = idx ;

var v = $prop('DynLeaderboardsPlugin.Dynamic.' + idx + '.Gap.Dynamic.ToFocused')
if (v == null) {
	return ''
}

if (v > 10000) {
	return format(v - 100000, '0', true) + 'L'
}

return format(v, '0.00', true)`

Concept show in attached picture
![image](https://user-images.githubusercontent.com/42914986/235353531-4ae56abd-13c4-45c3-a811-129cca110412.png)

**Comment by kaiusl** [comment]
The idea is correct (also indexes) but the gap returned by `Gap.Dynamic.ToFocused` changes based on the selected leaderboard. As stated in the [wiki](https://github.com/kaiusl/KLPlugins.DynLeaderboards/wiki/Available-properties#gaps) it shows the following gaps:

> Gap that changes based on the currently selected dynamic leaderboard.
>
>    Overall -> Total gap to the leader.
>    Class -> Total gap to the class leader.
>    (Partial)RelativeOverall, (Partial)RelativeClass -> Total gap to the focused car.
>    RelativeOnTrack -> Relative on track gap to the focused car.

So your code is only correct if any of the Relative leaderboards is selected. However if (for example) Overall leaderboard is selected it would show the gaps to the leader, not to you. I believe what you actually want is `Gap.ToFocused.Total` to always get the gap to you (or `Gap.ToFocused.OnTrack` if you want to get the on track gap to you).

So a working code is following
```javascript
var idx = $prop('DynLeaderboardsPlugin.Dynamic.FocusedPosInCurrentLeaderboard')
var ahead = $prop('DynLeaderboardsPlugin.Dynamic.' + idx + '.Gap.ToFocused.Total')
if (ahead == null) {
    return ''
}

if (ahead > 10000) {
    return format(ahead - 100000, '0', true) + 'L'
}

return format(ahead, '0.00', true)
``` 
```javascript
var idx = $prop('DynLeaderboardsPlugin.Dynamic.FocusedPosInCurrentLeaderboard') + 2
var behind = $prop('DynLeaderboardsPlugin.Dynamic.' + idx + '.Gap.ToFocused.Total')
// same formatting code
``` 
Note that the gap to ahead is always positive and gap to behind is negative. If you want other signs you can easily correct it. 

This version also allows to change which cars are the ahead and behind car. For example if you have selected Overall/RelativeOverall/PartialRelativeOverall leaderboard then the cars ahead and behind are in overall order. If Class/... leaderboard then the cars ahead and behind are the ones in class and so on.

If you only want the gap to the car ahead/behind in overall order then there is also `Gap.ToAhead.Overall` property. Note however that then indexes are slightly different and the gaps are always positive
```javascript
var idx = $prop('DynLeaderboardsPlugin.Dynamic.FocusedPosInCurrentLeaderboard') + 1
var ahead = $prop('DynLeaderboardsPlugin.Dynamic.' + idx + '.Gap.ToAhead.Overall')
// same formatting code
``` 
```javascript
var idx = $prop('DynLeaderboardsPlugin.Dynamic.FocusedPosInCurrentLeaderboard') + 2
var behind = $prop('DynLeaderboardsPlugin.Dynamic.' + idx + '.Gap.ToAhead.Overall')
// same formatting code
```

**Reply by kaiusl** [reply]
Just noticed that above suggestion with `Gap.ToFocused.Total` would show also total gap if RelativeOnTrack leaderboard was selected. So if the car ahead was lap down it would show -1L, which probably is not what one would expect.

And actually there is another property that I think fits perfectly for this usecase: `Gap.Dynamic.ToAhead`, which shows the following gaps 

> Gap to the car ahead that changes based on the currently selected dynamic leaderboard.
>
>   (Partial)(Relative)Overall -> Total gap to the car ahead in overall order.
>   (Partial)(Relative)Class -> Total gap to the car ahead in class.
>   RelativeOnTrack -> Relative gap to the car ahead on track.

So I would go with

```javascript
var idx = $prop('DynLeaderboardsPlugin.Dynamic.FocusedPosInCurrentLeaderboard') + 1
var ahead = $prop('DynLeaderboardsPlugin.Dynamic.' + idx + '.Gap.Dynamic.ToAhead')
// same formatting code
```
```javascript
var idx = $prop('DynLeaderboardsPlugin.Dynamic.FocusedPosInCurrentLeaderboard') + 2
var behind = $prop('DynLeaderboardsPlugin.Dynamic.' + idx + '.Gap.Dynamic.ToAhead')
// same formatting code
```
Which correctly chooses the car ahead/behind and expected gap based on the selected leaderboard type.

Note that to get the lap times of car ahead/behind you need to use just `var idx = $prop('DynLeaderboardsPlugin.Dynamic.FocusedPosInCurrentLeaderboard')` (without +1) for ahead.

```javascript
var idx = $prop('DynLeaderboardsPlugin.Dynamic.FocusedPosInCurrentLeaderboard')
var ahead = $prop('DynLeaderboardsPlugin.Dynamic.' + idx + '.Laps.Last.Time')
```
```javascript
var idx = $prop('DynLeaderboardsPlugin.Dynamic.FocusedPosInCurrentLeaderboard') + 2
var behind = $prop('DynLeaderboardsPlugin.Dynamic.' + idx + '.Laps.Last.Time')
```
